runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPipelineFuser.java

```diff
@@ -0,0 +1,222 @@
+                      !consumedCollectionsAndTransforms.containsKey(collectionConsumer))
+              .collect(Collectors.toSet());
+      checkState(
+          siblingSet.equals(sibs) || siblingSet.isEmpty(),
+          "Inconsistent collection of siblings reported for a %s. Initial attempt missed %s",
+          PCollectionNode.class.getSimpleName(),
+          Sets.difference(sibs, siblingSet));
+      if (siblingSet.isEmpty()) {
+        LOG.debug("Filtered out duplicate stage root {}", sibs);
+        continue;
+      }
+      ExecutableStage stage = fuseSiblings(siblingSet);
+      for (CollectionConsumer sibling : siblingSet) {
+        ExecutableStage oldStage = consumedCollectionsAndTransforms.put(sibling, stage);
+        // This should never happen; we filter out all of the existing consumers
+        checkState(
+            oldStage == null,
+            "Multiple %s registered for %s %s: %s and %s",
+            ExecutableStage.class.getSimpleName(),
+            CollectionConsumer.class.getSimpleName(),
+            sibling,
+            stage,
+            oldStage);
+      }
+      stages.add(stage);
+      for (PCollectionNode materializedOutput : stage.getOutputPCollections()) {
+        Set<CollectionConsumer> materializedConsumers =
+            getDownstreamInEnvConsumers(materializedOutput);
+        Collection<Set<CollectionConsumer>> siblings = groupSiblings(materializedConsumers);
+        siblingSets.addAll(siblings);
+      }
+    }
+    // TODO: Stages can be fused with each other, if doing so does not introduce duplicate paths
+    // for an element to take through the Pipeline. Compatible siblings can generally be fused,
+    // as can compatible producers/consumers if a PCollection is only materialized once.
+  }
+
+  private Collection<Set<CollectionConsumer>> getRootEnvTransforms(PTransformNode rootNode) {
+    checkArgument(
+        rootNode.getTransform().getInputsCount() == 0,
+        "%s is not at the root of the graph (consumes %s)",
+        PTransformNode.class.getSimpleName(),
+        rootNode.getTransform().getInputsMap());
+    if (pipeline.getEnvironment(rootNode).isPresent()) {
+      throw new IllegalArgumentException(
+          String.format(
+              "%s requires all root nodes to be runner-implemented %s primitives",
+              GreedyPipelineFuser.class.getSimpleName(),
+              PTransformTranslation.IMPULSE_TRANSFORM_URN));
+    } else {
+      unfusedTransforms.add(rootNode);
+      Set<CollectionConsumer> environmentNodes = new HashSet<>();
+      // Walk down until the first environments are found, and fuse them as appropriate.
+      for (PCollectionNode output : pipeline.getOutputPCollections(rootNode)) {
+        environmentNodes.addAll(getDownstreamInEnvConsumers(output));
+      }
+      return groupSiblings(environmentNodes);
+    }
+  }
+
+  private Set<CollectionConsumer> getDownstreamInEnvConsumers(PCollectionNode inputPCollection) {
```

Comment:
Try to stick with predecessor/successor/ancestor/descendant.

`getDescendantConsumersInEnv`